src/components/Video.h

```diff
@@ -25,13 +25,39 @@ along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
 
 #include <SDL_opengl.h>
 
+class VideoContext
```

Comment:
Can't all this be just a function pointer?

Reply:
It could, and I get that this might look overkill, but I decided on this approach for some reasons:

* A function pointer cannot hold state, so I'd have to put the window handle on a global variable to access it. Using a callable object (a lambda in this case) the window handle can be captured and accessed without hassle.

* The video context object can be extended in the future for other things. RetroArch has a similar structure, `video_frame_info_t`, that besides the swap buffers callback contains a bunch of other configurations and callbacks that the video driver can use to access stuff outside its control. As an example, an on-screen fps display could use a callback in the video context object to obtain the current fps value.

* The callback is always the same, and a template parameter is the best way to encode static polymorphism. Not that a single virtual call per frame would be noticeable, but whilst it might look less efficient, even with the `unique_ptr` and `virtual` calls, the compiler can actually inline the call, because everything is known at compile time: https://godbolt.org/z/r7gEwV

Reply:
> A function pointer cannot hold state, so I'd have to put the window handle on a global variable to access it. Using a callable object (a lambda in this case) the window handle can be captured and accessed without hassle.

You can use lambda captures in function pointers too. Just us a `std::function<void()>`. Here's an example in RAIntegration: <https://github.com/RetroAchievements/RAIntegration/blob/8937432e7496365cedafd165c53b9905bbd522fb/src/data/EmulatorContext.hh#L128>
This illustrates storing the function pointer. The actual caller doesn't provide any lambda state, but it could.